Feat/custom onDropHandler for viewportGrid

Jaseel-trenser opened the pull request:
<!-- Do Not Delete This! pr_template -->
<!-- Please read our Rules of Conduct: https://github.com/OHIF/Viewers/blob/master/CODE_OF_CONDUCT.md -->
<!-- 🕮 Read our guide about our Contributing Guide here https://docs.ohif.org/development/contributing -->
<!-- :hand: Thank you for starting this amazing contribution! -->

<!--
⚠️⚠️ Please make sure the checklist section below is complete before submitting your PR.
To complete the checklist, add an 'x' to each item: [] -> [x]
(PRs that do not have all the checkboxes marked will not be approved)
-->

### Context
- An option has been implemented to allow custom onDropHandler in viewort grid. This enhancement enables users for handling of additional functionalities during the onDrop event to the viewport.
- This PR has been incorporated by [FlyWheel.io](https://flywheel.io/).
<!--
Provide a clear explanation of the reasoning behind this change, such as:
- A link to the issue being addressed, using the format "Fixes #ISSUE_NUMBER"
- An image showing the issue or problem being addressed (if not already in the issue)
- Error logs or callStacks to help with the understanding of the problem (if not already in the issue)
-->

### Changes & Results
- Introduced a custom onDropHandler to handle additional functionalities during the onDrop to the viewpor
<!--
List all the changes that have been done, such as:
- Add new components
- Remove old components
- Update dependencies

What are the effects of this change?
- Before vs After
- Screenshots / GIFs / Videos
-->

### Testing

<!--
Describe how we can test your changes.
- open a URL
- visit a page
- click on a button
- etc.
-->

### Checklist

#### PR

<!--
https://semantic-release.gitbook.io/semantic-release/#how-does-it-work

Examples:
Please note the letter casing in the provided examples (upper or lower).

- feat(MeasurementService): add ...
- fix(Toolbar): fix ...
- docs(Readme): update ...
- style(Whitespace): fix ...
- refactor(ExtensionManager): ...
- test(HangingProtocol): Add test ...
- chore(git): update ...
- perf(VolumeLoader): ...

You don't need to have each commit within the Pull Request follow the rule,
but the PR title must comply with it, as it will be used as the commit message
after the commits are squashed.
-->

- [x] My Pull Request title is descriptive, accurate and follows the
  semantic-release format and guidelines.

#### Code

- [x] My code has been well-documented (function documentation, inline comments,
  etc.)

#### Public Documentation Updates

<!-- https://docs.ohif.org/ -->

- [x] The documentation page has been updated as necessary for any public API
  additions or removals.

#### Tested Environment

- [x] OS: Windows 11<!--[e.g. Windows 10, macOS 10.15.4]-->
- [x] Node version: 18.20.3 <!--[e.g. 18.16.1]-->
- [x] Browser: Chrome
  <!--[e.g. Chrome 83.0.4103.116, Firefox 77.0.1, Safari 13.1.1]-->

<!-- prettier-ignore-start -->
[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch
<!-- prettier-ignore-end -->

sedghi commented:
Can you rebase on top of the [new customization service](https://github.com/OHIF/Viewers/pull/4688) and follow what's done there please? Sorry for the inconvenience

Jaseel-trenser commented:
> Can you rebase on top of the [new customization service](https://github.com/OHIF/Viewers/pull/4688) and follow what's done there please? Sorry for the inconvenience

@sedghi I’ve rebased and updated the changes based on the new customization service.

sedghi commented on platform/app/src/components/ViewportGrid.tsx:
```diff
@@ -208,8 +213,24 @@ function ViewerViewportGrid(props: withAppTypes) {
-    const updatedViewports = _getUpdatedViewports(viewportId, displaySetInstanceUID);
-    viewportGridService.setDisplaySetsForViewports(updatedViewports);
+    let dropHandlerPromise = Promise.resolve({ handled: false });
+    const customOnDropHandler = customizationService.getCustomization('customOnDropHandler');
```
you need to provide a default null in the customizations folder in the extension. Also document it in here https://docs.ohif.org/platform/services/customization-service/customizationService#customizable-parts-of-ohif

sedghi commented on platform/app/src/components/ViewportGrid.tsx:
```diff
@@ -208,8 +213,20 @@ function ViewerViewportGrid(props: withAppTypes) {
-    const updatedViewports = _getUpdatedViewports(viewportId, displaySetInstanceUID);
-    viewportGridService.setDisplaySetsForViewports(updatedViewports);
+    const customOnDropHandler = customizationService.getCustomization('customOnDropHandler');
+    const dropHandlerPromise = customOnDropHandler({
+      ...props,
+      viewportId,
+      displaySetInstanceUID,
+      appConfig,
+    });
+
+    dropHandlerPromise.then(({ handled }) => {
+      if (!handled) {
+        const updatedViewports = _getUpdatedViewports(viewportId, displaySetInstanceUID);
+        viewportGridService.setDisplaySetsForViewports(updatedViewports);
+      }
+    });
```
Does it need to be async?

sen-trenser replied:
@sedghi As drag and drop a series may require additional inputs from user, it seems to be a good candidate for async and do the rest of the viewport updates if not handled by the custom drop handler.

Some example scenarios:
1. The user may be prompted to save or discard any changes (measurement/segmentation/other custom panel data) before updating viewports. If user accidentally dropped the series and want to cancel the action, the rest of drop handling code can be skipped after we get an async input from user. 
2. The user may be prompted to choose a view for the data like frame view, or select a sort order for 4D data (by trigger time, or patient position etc.) before displaying it in viewports.
3. Some async data need to be lazily fetched from API services when a series is dropped and before updating viewport. Although the data could be pre-loaded, the fetching could be async to ensure data is available if user drops data just after study loads.

sedghi replied:
Sure it makes sense, can you resolve my other comment and we are good to go
Thanks

Jaseel-trenser replied:
I have resolved the comments. Thanks

sedghi commented on extensions/default/src/customizations/onDropHandlerCustomization.ts:
```diff
@@ -0,0 +1,7 @@
+export default {
+  customOnDropHandler: {
+    $set: props => {
+      return Promise.resolve({ handled: false });
+    },
+  },
```
you don't need a $set here

Jaseel-trenser replied:
Updated

sedghi reviewed (approved):
Thanks!
